### DISCUSSION TITLE
AaC Docstring Convention

### DISCUSSION THREAD

**Discussion by jondavid-black** [DevOps-MBSE/AaC#49] [Archive] [answered]
In a recent PR review there was a suggestion to standardize our usage of python docstrings.

When I started adding docstrings I tried to follow the guidance provided here:  https://realpython.com/documenting-python-code/

I'm open to suggestions on what works best for the team...and importantly the down-stream stakeholder building a custom extension or plugin.

What do you recommend we do and why?

**Comment by Coffee2Bits** [comment]
I can't find a python standard that dictates params/returns in the docstring conventions standard, [pep-257](https://www.python.org/dev/peps/pep-0257/). So, if we want to implement docstring standards that dictate args/returns in docstrings then I think we should consider the `numpy` and `google` standards, especially since flake8 supports docstring linting for those two conventions. I've seen some uses of `:param:`, `:type:`, and `:return:`, but those look like they're for use with a documentation tool, [Sphinx](https://www.sphinx-doc.org/en/master/usage/restructuredtext/domains.html#info-field-lists).

[Numpy Style Documentation](https://numpydoc.readthedocs.io/en/latest/format.html#docstring-standard)
[Google Style Documentation](https://google.github.io/styleguide/pyguide.html#38-comments-and-docstrings)

I'm not particularly opinionated about which docstring standard we should use, numpy and google are pretty similar. If I had to cast a vote right now, I'd vote for Google because it's a dash easier to read than numpy.

**Comment by jondavid-black** [comment]
It looks like there are a couple of popular ways to generate documentation for your python project.

- [Sphinx](https://www.sphinx-doc.org/en/master/index.html):  Described as the most popular and most powerful in the articles I've read.  Also described as complex.
- [Pdoc](https://github.com/mitmproxy/pdoc):  Described as the 2nd most popular, a bit less powerful, and very easy to adopt.

My vote is to do with the KISS (Keep It Simple Stupid) approach here and just use pdoc.  I like that it supports markdown styling in docstrings.  I'll do a little experiment on the side with pdoc and different styles to see what comes out and report back.

So that addresses tooling, but formatting is another matter.  I'm with Alex...I like the Google Style.  I think it's just a tiny bit more readable and writable than Numpy...and a bit more concise.

I'll report back once I have run an experiment and we'll settle on something based on the outcome.

**Reply by cvchaparro** [reply]
I second keeping everything simple and readable (especially when it's concise). My vote is to go with Google Style and Pdoc.

**Comment by jondavid-black** [comment]
I've started some experimentation.  So far I'd say Google Style works best with VSCode hinting.  Numpy Style is second and RsT Style is third.  Pdoc doesn't seem to work well with Google Style or RsT Style but works great with Numpy Style.  I've still got to do some testing with Sphinx.  More to follow.

As for "in code" readability, I think Google and Numpy Style are equivalent, but RsT style is too complex in its notation.

At this point I'd say Google Style and Numpy Style are tied with RsT Style a distant third.  This is strange because I often read that RsT Style is the "preferred standard".  I'm a little worried I may be missing something.  Maybe that will be revealed when I start testing with Sphinx.

**Comment by jondavid-black** [comment, marked as answer]
I finished my simple experiment and [published the results](https://github.com/jondavid-black/pydocstrtest).

Since we're concerned with approachability we have a bit of a conflict.  We want very nice documentation but we also want very nice developer experience.  I think the best of both world for us is to use Google style docstrings and Sphinx to generate documentation.  It's a bit of a mix of the opinion reference above.  I think we can take on the setup and complexity of running and maintaining Sphinx for the core project, but we also provide clean hinting in VSCode.

If you guys agree I'll write an issue to setup sphinx documentation.  I guess we need to start a "How to Contribute" page as well to document out decisions for others.

**Reply by Coffee2Bits** [reply]
There is a Sphinx plugin for numpy/google style docstrings called napoleon. Here is the documentation for it: https://www.sphinx-doc.org/en/master/usage/extensions/napoleon.html

**Reply by jondavid-black** [reply]
I tested that plugin...totally works.  Now each of the styles renders exactly the same.

**Reply by Coffee2Bits** [reply]
Fantastic! Since that's the case, my vote for docstring style is the google style.

**Reply by cvchaparro** [reply]
I concur - let's go with Google style.

**Comment by lizzcondrey** [comment]
Since this was marked resolved, answered selected, the standard has been adhered to, and the docstring support with Sphinx is in place, I am going to close this discussion.
